tests/Optimizer/Transformer/PreloadHeroImageTest.php

```diff
@@ -205,6 +205,55 @@ public function dataTransform()
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero1.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero2.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero3.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero4.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero5.png"></amp-img>'
+                ),
+                $output(
+                    '<amp-img width="500" height="400" src="/foo.png"></amp-img>'
+                    . '<amp-img data-hero data-hero-candidate width="500" height="400" src="/hero1.png" i-amphtml-ssr><img class="i-amphtml-fill-content i-amphtml-replaced-content" decoding="async" src="/hero1.png"></amp-img>'
+                    . '<amp-img data-hero data-hero-candidate width="500" height="400" src="/hero2.png" i-amphtml-ssr><img class="i-amphtml-fill-content i-amphtml-replaced-content" decoding="async" src="/hero2.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero3.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero4.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero5.png"></amp-img>'
+                ),
+            ],
+
+            'hero images trump hero image candidates' => [
+                $input(
+                    '<amp-img width="500" height="400" src="/foo.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero1.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero2.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero3.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero4.png"></amp-img>'
+                    . '<amp-img data-hero width="500" height="400" src="/hero5.png"></amp-img>'
+                ),
+                $output(
+                    '<amp-img width="500" height="400" src="/foo.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero1.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero2.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero3.png"></amp-img>'
+                    . '<amp-img data-hero-candidate width="500" height="400" src="/hero4.png"></amp-img>'
+                    . '<amp-img data-hero width="500" height="400" src="/hero5.png" i-amphtml-ssr><img class="i-amphtml-fill-content i-amphtml-replaced-content" decoding="async" src="/hero5.png"></amp-img>'
+                ),
+            ],
```

Comment:
Can tests be added for fallbacks as well as testing the placeholder fallback image deduplication logic?

Reply:
Those are already tested. The fallbacks are tested in the first 2 tests (`detects regular image as hero image` & `skips tiny images`), for example. The placeholders are tested with the tests that are using placeholders, like `fetches placeholders for animations`, as these would otherwise have the SSR'ed `<img>` element twice (which was what detected the bug in the first place).